letsencrypt/cli.py

```diff
@@ -304,31 +304,91 @@ def _auth_from_domains(le_client, config, domains, plugins):
 
     return lineage
 
+class ConfiguratorError(TypeError): # pylint: disable=missing-docstring
+    pass
+
+def set_configurator(previously, now):
+    """Setting configurators multiple ways is okay, as long as they all agree"""
+    if now is None:
+        # we're not actually setting anything
+        return previously
+    if previously:
+        if previously != now:
+            msg = "Too many flags setting configurators/installers/authenticators %s -> %s"
+            raise ConfiguratorError, msg % (`previously`, `now`)
+    return now
+
+def diagnose_configurator_problem(cfg_type, requested):
+    """
+    Raise the most helpful error message about a plugin being unavailable
+    @cfg_type -- "installer" or "authenticator"
+    @requested -- which plugin the user wanted
+    """
+    plugins = plugins_disco.PluginsRegistry.find_all()
+
+    if requested:
+        if requested not in plugins:
+            msg = "The requested " + requested + " plugin does not appear to be installed"
+            raise ConfiguratorError, msg
+        else:
+            msg = "The " + requested + " plugin is not working; there may be problems "
+            msg += "with your existing configuration"
+            raise ConfiguratorError, msg
+    raise ConfiguratorError, cfg_type + " could not be determined or is not installed"
+
+
+
+def choose_configurator_plugins(args, config, plugins, verb):
+    """Figure out which configurator we're going to use"""
+
+    # Which plugins do we need?
+    need_inst = need_auth = (verb == "run")
+    if verb == "auth":
+        need_auth = True
+    if verb == "install":
+        need_inst = True
+
+    # Which plugins did the user request?
+    req_inst = req_auth = args.configurator
+    req_inst = set_configurator(req_inst, args.installer)
+    req_auth = set_configurator(req_auth, args.authenticator)
+    if args.nginx:
+        req_inst = set_configurator(req_inst, "nginx")
+        req_auth = set_configurator(req_auth, "nginx")
+    if args.apache:
+        req_inst = set_configurator(req_inst, "apache")
+        req_auth = set_configurator(req_auth, "apache")
+
+    logger.debug("Requested authenticator %s and installer %s", req_auth, req_inst)
+
+    # Try to meet the user's request and/or ask them to pick plugins
+    if verb == "run" and req_auth == req_inst:
+        # Unless the user has explicitly asked for different auth/install,
+        # only consider offering a single choice
+        authenticator = installer = display_ops.pick_configurator(config, req_inst, plugins)
+    else:
+        if need_inst:
+            installer = display_ops.pick_installer(config, req_inst, plugins)
```

Comment:
So in the scenario described in #896, this will pick an installer if one is installed. Suppose Apache is installed. The behavior is then equivalent to `letsencrypt -a standalone -i apache` which will almost certainly not work due to bound ports. Is this the behavior we want?